Vault general fixes and security hardening

bcoca opened the pull request:
Set new default Vault cipher to take the place of the, now deemed week, AES256 cipher, which is still kept for backwards compatibility.

also fixes #51862

##### ISSUE TYPE

- Bugfix Pull Request

bcoca committed:
Vault loadable ciphers and fixes

To deal with issues with current AES256 cipher we
changed a bit how ciphers are loaded to enabled a
simpler way to update/upgrade ciphers going forward

  * We now validate vault parameters to avoid injection
  * Added new config to set vault cipher used for encryption
  * Ciphers are now used as classes instead of objects
  * Implemented 'cipher loader' to encapsulate and simplify
    dealing with mulitple ciphers
  * New ciphers added, kept aes256 for backwards compat
  * New ciphers used base64 armor over hexlify
  * Updated tests to be more cipher agonstic and/or use dummy cipher

webknjaz commented:
/azp run

webknjaz commented:
/azp run

webknjaz commented on test/integration/targets/ansible-vault/runme.sh:
```diff
@@ -390,11 +380,14 @@ echo "The stdout log had 'New Vault password' in it and it is not supposed to. r
-head -1 "${TEST_FILE_EDIT2}" | grep "${FORMAT_1_2_HEADER};vault_password"
+# shell check ignored as $ANSIBLE_VAULT not a shell var
+# shellcheck disable=SC2016
+head -1 "${TEST_FILE_EDIT2}" | grep -E '$ANSIBLE_VAULT;1.2;.+;vault_password'
```
Shouldn't `$` be escape in this regex, though? Just like it used to be in `$FORMAT_1_2_HEADER`. Otherwise, it'd be expecting EOL in that place IIUC.
```suggestion
head -1 "${TEST_FILE_EDIT2}" | grep -E '\$ANSIBLE_VAULT;1.2;.+;vault_password'
```

bcoca replied:
yep, fixed but not pushed as i was asked to not push while it was being reviewed

webknjaz commented on test/integration/targets/ansible-vault/runme.sh:
```diff
@@ -390,11 +380,14 @@ echo "The stdout log had 'New Vault password' in it and it is not supposed to. r
 # edit a 1.2 format with vault id, should keep vault id and 1.2 format
 EDITOR=./faux-editor.py ansible-vault edit "$@" --vault-id vault_password@vault-password "${TEST_FILE_EDIT2}"
-head -1 "${TEST_FILE_EDIT2}" | grep "${FORMAT_1_2_HEADER};vault_password"
+# shell check ignored as $ANSIBLE_VAULT not a shell var
+# shellcheck disable=SC2016
```
With `$` escaped, would this ignore still be necessary?

nitzmahone committed:
minimize new API surface area
* cut down cipher implementation size
* add basic testing framework
* fix tests

bcoca committed:
minor cleanup

ci_complete

nitzmahone committed:
move vault-id validation to format_vaulttext_envelope

* update/optimize tests
* minimize delta for backport

nitzmahone committed:
misc changes, rebased from current PR state

move vault-id validation to format_vaulttext_envelope

* update/optimize tests
* minimize delta for backport

nitzmahone committed:
move rot13 to test code, wrap access in a reusable fixture

bcoca committed:
move output to next line when un/vaulting to tty

bcoca committed:
updated clog

bcoca committed:
use config to verify passed in cipher name

currently should only affect filter, but should take care of other direct callers

bcoca committed:
renamed to method and v2

renamed in code, rename in tests, renamed in jest and even ignore rename

bcoca committed:
use updated var

bcoca committed:
rename in tests

bcoca committed:
add ro13 as choice when testing

handle method name fixes, moved data tests, renamed headers

nitzmahone committed:
misc test cleanups and coverage increase
* additional tests
* remove dead code

nitzmahone committed:
remove dead code

nitzmahone committed:
fix sanity tests
* make script-noshebang executable only at test runtime to avoid shebang sanity tests

nitzmahone committed:
temp restore iffy error handler

nitzmahone committed:
fix typing issues
* enable type checking on more public vault API

bcoca committed:
better desc

nitzmahone committed:
restore error detail

bcoca committed:
better error on bad method on encrypt

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -653,7 +653,7 @@ def decrypt_and_get_vault_id(self, vaulttext: str | bytes, filename: str | None
                 break
             except (ValueError, TypeError) as exc:
-                msg = u"There was a vault format error"
+                msg = f"There was a vault format error: {exc}"
```
should not need this since you are doing `from exc` later on or even adding orig_exc would do same

konstruktoid commented on changelogs/fragments/vault_ciphers.yml:
```diff
@@ -0,0 +1,11 @@
+bugfixes:
+  - Set new default Vault method to take the place of the, now deemed week, AES256 method.
```
week -> weak

bcoca committed:
fix errors, let them percolate to top

 can remove adding exc test as it would now be dupe

 ansible.parsing.vault.AnsibleVaultError: Unable to encrypt vault with unsupported method. Invalid value "rotten" for configuration option "setting: VAULT_METHOD ", valid values are: aes256, v2

mattclay committed:
Use salt and encryption key bytes directly

Also fix docstring.

mattclay commented on lib/ansible/config/base.yml:
```diff
@@ -1171,11 +1171,29 @@ DEFAULT_VAULT_IDENTITY:
+  default: v2
+  choices:
+    'aes256': AES 256 with PBKDF2HMAC key derivation with double hexlify byte shield
+    'v2': Using cryptography's fernet API with dual keys, AES 128, CBC and sCrypt key derivation, with base64 byte shield
```
We shouldn't need to explain *how* Fernet works. It's sufficient to state we're using it (along with Scrypt):

```suggestion
    'v2': Fernet with dual keys and Scrypt key derivation with a base64 byte shield
```

bcoca replied:
I'm just saving them a lookup by giving the info up front, also matching the other method

bcoca committed:
fix typos

bcoca committed:
fix test

bcoca committed:
better msg, also test it

mattclay committed:
Improve error handling for invalid vault method name

When config is used to specify the vault method, config reports the error.
When the CLI is used, argparse handles the error because choices is set.
When the API is used, load_vault_method reports the error.

This ensures the error message is appropriate for the context in which the invalid method was specified.

mattclay committed:
Fix integration test failure

bcoca committed:
simpler config access

also some fixes

mattclay committed:
Remove unused import

bcoca committed:
better envelope validation and tests

bcoca committed:
fixes

bcoca committed:
test choices

bcoca committed:
new method v2b

Also fix x2 exception reporting
Only caches on decrypt, encrypt won't hit cache due to salt generation
note: w/o cache, decrypt takes same time as encrypt

bcoca committed:
fix tests

bcoca committed:
really, fix tests

bcoca committed:
fix tab

bcoca committed:
gnore import issue

bcoca committed:
readded exc to msg

bcoca committed:
Key pairs method

bcoca committed:
allow both pem or ssh formatted keys

pb82 commented:
@bcoca how does this affect existing secrets encrypted with an older version of vault (and the old cipher)? Do I now need to specify `method_name` in order to decrypt them?

If so, I wonder if this would also require changes in Ansible Tower to handle existing secrets gracefully.

bcoca commented:
Decryption takes the info from the vault itself, so as not to impose any additional requirements on the user.  

This existing vaults will keep working until we fully retire the old encryption method in the 'far off' future, there will be a full deprecation cycle then. 

The new method becomes the default for encrypting in this PR, so if you want to continue using the old method to encrypt, then you do need to pass that explicitly, either on the CLI or via configuration.

There is a documentation update that will go out at the time this PR is merged to ensure all this is explained to users.

bcoca committed:
cleanup kpv1
